libtrap/src/ifc_tcpip.c

```diff
@@ -960,419 +958,533 @@ static int client_socket_connect(void *priv, const char *dest_addr, const char *
+ * \brief Write buffer size to its header and shift active index.
  *
- * * When everybody fails due to timeout, everything
- * is alright, everybody stays in 'current' buffer.
- *
- * * When somebody fails and somebody is successful, failing clients are moved into 'backup' buffer and they have chance to receive the rest of message later. We disconnect all clients that were in 'backup' buffer before moving new ones from 'current' buffer.
- *
- * * All successful clients from 'backup' buffer are moved back into 'current' buffer.
- *
- * \param[in] priv  pointer to module private data
- * \param[in] data  pointer to data to send
- * \param[in] size  size of data to send
- * \param[in] timeout  timeout in microseconds
- * \return 0 on success (TRAP_E_OK), TRAP_E_BAD_FPARAMS if sender was not properly initialized, TRAP_E_TERMINATED if interface was terminated.
+ * \param[in] priv Pointer to output interface private structure.
+ * \param[in] buffer Pointer to the buffer.
  */
-int tcpip_sender_send(void *priv, const void *data, uint32_t size, int timeout)
+static inline void finish_buffer(tcpip_sender_private_t *priv, buffer_t *buffer)
 {
-   uint8_t buffer[DEFAULT_MAX_DATA_LENGTH];
-   int result = TRAP_E_TIMEOUT;
-   tcpip_sender_private_t *c = (tcpip_sender_private_t *) priv;
-   /* timeout for select */
-   struct timeval tv;
-   /* timeout for sem_timedwait */
-   struct timespec ts = { .tv_sec = 0, .tv_nsec = 0 };
-   fd_set set, disset;
-   int maxsd = -1;
-   struct client_s *cl;
-   uint32_t i, j, failed, passed;
-   int retval;
-   ssize_t readbytes;
-   /* first timestamp for global timeout in this function...
-    * in the RESET state, we should check the timeout given by caller
-    * with elapsed time from entry_time.
-    * Timeout is in microseconds... */
+   uint32_t header = htonl(buffer->wr_index);
+   memcpy(buffer->header, &header, sizeof(header));
 
-   uint64_t entry_time = get_cur_timestamp();
-   uint64_t curr_time = 0, elapsed_time;
-   uint64_t send_entry_time;
-   uint64_t send_exit_time;
+   priv->active_buffer = (priv->active_buffer + 1) % priv->buffer_count;
+   priv->autoflush_timestamp = get_cur_timestamp();
 
-   char block = ((timeout == TRAP_WAIT || timeout == TRAP_HALFWAIT) ? 1 : 0);
+   buffer->clients_bit_arr = priv->clients_bit_arr;
+   buffer->wr_index = 0;
+}
 
-   /* pointer to timeout for select() */
-   struct timeval *tv_p = ((block != 0) ? NULL : &tv);
-   /* pointer to timeout for sem_timedwait() */
-   struct timespec *ts_p = &ts;
+/**
+ * \brief Force flush of active buffer
+ *
+ * \param[in] priv pointer to interface private data
+ */
+void tcpip_sender_flush(void *priv)
+{
+   tcpip_sender_private_t *c = (tcpip_sender_private_t *) priv;
+   c->autoflush_timestamp = get_cur_timestamp();
 
-   /* correct module will pass only possitive timeout or TRAP_WAIT, TRAP_HALFWAIT */
-   assert(timeout >= TRAP_HALFWAIT);
+   pthread_mutex_lock(&c->ctx->out_ifc_list[c->ifc_idx].ifc_mtx);
```

Comment:
If I look at the design, this lock should not be used inside functions itself, but should surround this function in trap.c when called. What do you think @cejkato2 ?

Reply:
Well, the best knowledge can provide @barnama1, but I guess it is because autoflush needs to use the same mutex and it is when an ordinary send is called...

Thread safety is a complex topic. It should work now but I think we need some automatic tests to be sure...